dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalBg.cpp

```diff
@@ -174,15 +175,96 @@ class LocalDMFileGcRemover final
     }
 };
 
+// A callback class for find out the DMFiles on S3 referenced by this store.
+class S3DMFileGcScanner final
+{
+private:
+    // !!! Warning !!!
+    // Should only keep a weak ref of storage path pool since
+    // this callback instance may still valid inside the PageStorage
+    // even after the DeltaMerge storage is shutdown or released.
+    std::weak_ptr<StoragePathPool> path_pool_weak_ref;
+
+public:
+    explicit S3DMFileGcScanner(std::weak_ptr<StoragePathPool> path_pool_)
+        : path_pool_weak_ref(std::move(path_pool_))
+    {}
+
+    ExternalPageCallbacks::PathAndIdsVec operator()()
+    {
+        ExternalPageCallbacks::PathAndIdsVec path_and_ids_vec;
+
+        // If the StoragePathPool is invalid or shutdown flag is set,
+        // meaning we call `scanner` after shutdowning or dropping the table,
+        // simply return an empty list is OK.
+        auto path_pool = path_pool_weak_ref.lock();
+        if (!path_pool || path_pool->isShutdown())
+            return path_and_ids_vec;
+
+        auto delegate = path_pool->getStableDiskDelegator();
+        auto local_page_ids = delegate.getAllRemoteDTFilesForGC();
+        // path here is useless, just pass empty string here.
+        path_and_ids_vec.emplace_back("", std::move(local_page_ids));
+        return path_and_ids_vec;
+    }
+};
+
+// A callback class for removing the reference to DMFiles on S3 from local delegator.
+// The actual GC of DMFiles on S3 is managed by other class.
+class S3DMFileGcRemover final
+{
+private:
+    // !!! Warning !!!
+    // Should only keep a weak ref of storage path pool since
+    // this callback instance may still valid inside the PageStorage
+    // even after the DeltaMerge storage is shutdown or released.
+    std::weak_ptr<StoragePathPool> path_pool_weak_ref;
+
+public:
+    explicit S3DMFileGcRemover(std::weak_ptr<StoragePathPool> path_pool_)
+        : path_pool_weak_ref(std::move(path_pool_))
+    {}
+
+    void operator()(const ExternalPageCallbacks::PathAndIdsVec & path_and_ids_vec, const std::set<PageIdU64> & valid_ids)
+    {
+        // If the StoragePathPool is invalid or shutdown flag is set,
+        // meaning we call `remover` after shutdowning or dropping the table,
+        // we must skip because the `valid_ids` is not reliable!
+        auto path_pool = path_pool_weak_ref.lock();
+        if (!path_pool || path_pool->isShutdown())
+            return;
+
+        auto delegate = path_pool->getStableDiskDelegator();
+        for (const auto & [path, ids] : path_and_ids_vec)
+        {
+            for (auto id : ids)
+            {
+                if (!valid_ids.count(id))
+                {
+                    delegate.removeRemoteDTFile(id);
```

Comment:
maybe add some logging about removing the remote dtfile is helpful

Reply:
Added.